impala/hiveserver2.py

```diff
@@ -342,26 +355,29 @@ def __iter__(self):
         return self
 
     def __next__(self):
+        if len(self._buffer) == 0:
+            rows = self._fetch_batch(columnar=False)
+            self._buffer.extend(rows)
+            if len(self._buffer) == 0:
+                log.debug('__next__: no more rows to fetch')
+                raise StopIteration
+
+        log.debug('__next__: popping row out of buffer')
+        return self._buffer.pop(0)
+
+    def _fetch_batch(self, columnar=False):
         if not self.has_result_set:
-            raise ProgrammingError(
-                "Trying to fetch results on an operation with no results.")
-        if len(self._buffer) > 0:
-            log.debug('__next__: popping row out of buffer')
-            return self._buffer.pop(0)
-        elif self._last_operation_active:
+            raise ProgrammingError("Trying to fetch results on an operation "
+                                   "with no results.")
+        if self._last_operation_active:
             # self._buffer is empty here and op is active: try to pull more
             # rows
-            log.debug('__next__: buffer empty and op is active => fetching '
+            log.debug('fetch: buffer empty and op is active => fetching '
```

Comment:
How do you know that the buffer is empty here?  If this call to `_fetch_batch` comes after there was a preceding call to `fetchone`, the buffer might have some records in it that would get skipped.